### PR TITLE
86 feature eldritch pivotportscan

### LINKED ISSUE #86: [feature] Eldritch pivot.portscan
Add port scan function to eldritch.

### PR DESCRIPTION
**Pull request by hulto**
#### What type of PR is this?

Add one of the following kinds:
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind eldritch-function

#### What this PR does / why we need it:
Implements the port scan function to realm.

#### Which issue(s) this PR fixes:
Fixes #86

### CODE REVIEW

File: implants/eldritch/src/pivot/port_scan_impl.rs [hunk 1]
```diff
@@ -1,5 +1,662 @@
...
+            let test_cidr =  vec!["127.0.0.1/32".to_string()];
+
+            let _scan_res = handle_port_scan(test_cidr, test_ports.clone(), String::from("tcp"), 5).await?;
+        }
+        Ok(())
+    }
+
+    // verify our non async call works and Dict return type.
+    #[test]
+    fn test_starlark_dict_from_interpreter() -> anyhow::Result<()>{
```
Comment (hulto) [inline implants/eldritch/src/pivot/port_scan_impl.rs:609]: Follow function naming scheme `test_portscan_`

File: implants/eldritch/src/pivot.rs [hunk 2]
```diff
@@ -65,9 +66,9 @@ fn methods(builder: &mut MethodsBuilder) {
         smb_exec_impl::smb_exec(target, port, username, password, hash, command)
     }     
     // May want these too: PSRemoting, WMI, WinRM
-    fn port_scan(this:  PivotLibrary, target_cidrs: Vec<String>, ports: Vec<i32>, portocol: String) ->  anyhow::Result<Vec<String>> {
+    fn port_scan<'v>(this:  PivotLibrary, starlark_heap: &'v Heap, target_cidrs: Vec<String>, ports: Vec<i32>, portocol: String, timeout:  i32) ->  anyhow::Result<Vec<Dict<'v>>> {
```
Comment (Cictrone) [inline implants/eldritch/src/pivot.rs:69]: double check 'v is necessary

File: implants/eldritch/src/pivot/port_scan_impl.rs [hunk 3]
```diff
@@ -1,5 +1,662 @@
...
+use tokio::time::{Duration,sleep};
+use tokio::net::{TcpStream, UdpSocket};
+use async_recursion::async_recursion;
+
+macro_rules! scanf {
+    ( $string:expr, $sep:expr, $( $x:ty ),+ ) => {{
+        let mut iter = $string.split($sep);
+        ($(iter.next().and_then(|word| word.parse::<$x>().ok()),)*)
+    }}
+}
```
Comment (Cictrone) [inline implants/eldritch/src/pivot/port_scan_impl.rs:18]: pls no

File: implants/eldritch/src/pivot/port_scan_impl.rs [hunk 4]
```diff
@@ -1,5 +1,662 @@
...
+            result.push(int_to_string(host_u32).unwrap());
+        }
+
+        // Handle weird /31 cidr edge case
+        if host_u32 == (broadcast_u32 - 1) {
+            host_u32 = host_u32 + 1;
+            result.push(int_to_string(host_u32).unwrap());
+        }
+
+        // boardcast_u32-1 will not add the broadcast address for the net. Eg. 255.
```
Comment (Cictrone) [inline implants/eldritch/src/pivot/port_scan_impl.rs:113]: nit: boardcast

File: implants/eldritch/src/pivot/port_scan_impl.rs [hunk 5]
```diff
@@ -1,5 +1,662 @@
...
+    match response {
+        Ok(results) => {
+            let mut final_res: Vec<Dict> = Vec::new();
+            for row in results {
+                // Define underlying datastructure.
+                let res: SmallMap<Value, Value> = SmallMap::new();
+                // Create Dict type.
+                let mut tmp_res = Dict::new(res);
+
+                let tmp_value1 = starlark_heap.alloc_str(row.0.as_str());
```
Comment (Cictrone) [inline implants/eldritch/src/pivot/port_scan_impl.rs:350]: try using [] again

File: implants/eldritch/src/pivot/port_scan_impl.rs (hunk 1 above)
Comment (Cictrone) [reply implants/eldritch/src/pivot/port_scan_impl.rs:609]: +1

File: implants/eldritch/src/pivot/port_scan_impl.rs [hunk 6]
```diff
@@ -1,5 +1,662 @@
...
+        }
+
+        let globals = GlobalsBuilder::extended().with(func_port_scan).build();
+        let module: Module = Module::new();
+
+        let mut eval: Evaluator = Evaluator::new(&module);
+        let res: Value = eval.eval_module(ast, &globals).unwrap();
+
+        // println!("{:?}", res.to_string());
+        let expected_output = format!("[{{\"ip\": \"127.0.0.1\", \"port\": {}, \"protocol\": \"tcp\", \"status\": \"closed\"}}, {{\"ip\": \"127.0.0.1\", \"port\": {}, \"protocol\": \"tcp\", \"status\": \"closed\"}}, {{\"ip\": \"127.0.0.1\", \"port\": {}, \"protocol\": \"tcp\", \"status\": \"closed\"}}, {{\"ip\": \"127.0.0.1\", \"port\": {}, \"protocol\": \"tcp\", \"status\": \"closed\"}}]", test_ports[0], test_ports[1], test_ports[2], test_ports[3]);
```
Comment (Cictrone) [inline implants/eldritch/src/pivot/port_scan_impl.rs:655]: use a raw string to make this look not poopy

**Review by Cictrone — COMMENTED** [review]
lmk when all checked then i accept

File: implants/eldritch/src/pivot.rs (hunk 2 above)
Comment (hulto) [reply implants/eldritch/src/pivot.rs:69]: Nope it's a part of starlark auto heap magic.

File: implants/eldritch/src/pivot/port_scan_impl.rs (hunk 5 above)
Comment (hulto) [reply implants/eldritch/src/pivot/port_scan_impl.rs:350]: These are tuples not vecs :(

**Review by Cictrone — APPROVED** [review]

**Pull request merged**